internal/backend/runtime/omni/controllers/omni/machine_status.go

```diff
@@ -514,13 +515,26 @@ func (ctrl *MachineStatusController) handleNotification(ctx context.Context, r c
 				}
 			}
 
-			if spec.Schematic.Invalid {
-				spec.Schematic.InitialSchematic = ""
-			} else if spec.Schematic.InitialSchematic == "" {
+			if spec.Schematic.InitialSchematic == "" {
 				spec.Schematic.InitialSchematic = spec.Schematic.FullId
 			}
 
 			spec.Schematic.InAgentMode = event.Schematic.InAgentMode
+
+			// We can populate the initial set of extensions and kernel args only if the current schematic ID matches the initial schematic ID.
+			// Otherwise, we missed the moment to capture them.
+			if spec.Schematic.InitialState == nil && !spec.Schematic.Invalid && spec.Schematic.FullId == spec.Schematic.InitialSchematic {
+				spec.Schematic.InitialState = &specs.MachineStatusSpec_Schematic_InitialState{
+					Extensions: event.Schematic.Extensions,
+					KernelArgs: event.Schematic.KernelArgs,
+				}
+			}
+
+			// if the schematic is invalid or the machine is in agent mode, we reset the initial schematic information
+			if spec.Schematic.Invalid || spec.Schematic.InAgentMode {
```

Comment:
About the `inAgentMode` check, see https://github.com/siderolabs/omni/pull/1501/files#r2315640127.